base/array.jl

```diff
@@ -2436,6 +2436,48 @@ function filter!(f, a::AbstractVector)
     return a
 end
 
+"""
+     mask!(a::AbstractVector, m::AbstractVector{Bool})
+
+The inplace version of logical indexing `a = a[m]`. That is, `mask!(a, m)` on
+vectors of equal length `a` and `m` will remove all elements from `a` for which
+`m` at the corresponding index is `false`.
+
+# Examples
+```jldoctest
+julia> a = [:a, :b, :c];
+
+julia> mask!(a, [true, false, true])
+2-element Vector{Symbol}:
+ :a
+ :c
+
+julia> a
+2-element Vector{Symbol}:
+ :a
+ :c
+```
+"""
+function mask!(a::AbstractVector, m::AbstractVector{Bool})
+    j = firstindex(a)
+    for i in eachindex(a, m)
+        @inbounds begin
+            ai = a[i]
+            mi = m[i]
+            a[j] = ai
+        end
+        j = ifelse(mi, nextind(a, j), j)
```

Comment:
nit: there might be non-zero cost to copying an element, which the `ifelse` function implies (and I don't believe there's any simd optimization possible here due to the loop carried dependencies)
```suggestion
            if m[i]
                i == j || (a[j] = a[i])
                j = nextind(a, j)
            end
        end
```

Reply:
Interesting, should `filter!` also get an analogous change then? https://github.com/JuliaLang/julia/blob/master/base/array.jl#L2441

Reply:
Ah, interesting. yeah, seems like that could be a good change. This has already crossed to https://github.com/JuliaLang/julia/pull/36229 also. And might have some implications for https://github.com/JuliaArrays/OffsetArrays.jl/pull/137 already too?